cmd/lotus-shed/hello.go

```diff
@@ -0,0 +1,164 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/filecoin-project/go-address"
+	cborutil "github.com/filecoin-project/go-cbor-util"
+	"github.com/filecoin-project/lotus/blockstore"
+	"github.com/filecoin-project/lotus/chain/consensus/filcns"
+	"github.com/filecoin-project/lotus/chain/store"
+	"github.com/filecoin-project/lotus/chain/types"
+	lcli "github.com/filecoin-project/lotus/cli"
+	"github.com/filecoin-project/lotus/lib/addrutil"
+	"github.com/filecoin-project/lotus/node/hello"
+	"github.com/ipfs/go-datastore"
+	"github.com/libp2p/go-libp2p"
+	"github.com/libp2p/go-libp2p-core/crypto"
+	"github.com/libp2p/go-libp2p-core/host"
+	inet "github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/multiformats/go-multiaddr"
+	"github.com/urfave/cli/v2"
+	"golang.org/x/xerrors"
+)
+
+var helloCmd = &cli.Command{
+	Name:        "hello",
+	Description: "Get remote peer hello message by multiaddr",
+	Flags: []cli.Flag{
+		&cli.StringFlag{
+			Name:     "multiaddr",
+			Usage:    "remote peer multiaddr",
+			Required: true,
+		},
+		&cli.StringFlag{
+			Name:  "genesis-path",
+			Usage: "genesis car file path",
+		},
+	},
+
+	Action: func(cctx *cli.Context) error {
+		ctx := lcli.ReqContext(cctx)
+		cf := cctx.String("genesis-path")
+		f, err := os.OpenFile(cf, os.O_RDONLY, 0664)
+		if err != nil {
+			return xerrors.Errorf("opening the car file: %w", err)
+		}
+		bs := blockstore.FromDatastore(datastore.NewMapDatastore())
+		cs := store.NewChainStore(bs, bs, datastore.NewMapDatastore(), filcns.Weight, nil)
+		ts, err := cs.Import(cctx.Context, f)
+		if err != nil {
+			return err
+		}
+		pis, err := addrInfoFromArg(ctx, cctx)
+		if err != nil {
+			return err
+		}
+		priv, _, err := crypto.GenerateKeyPair(crypto.RSA, 2048)
+		if err != nil {
+			return err
+		}
+		h, err := libp2p.New(libp2p.Identity(priv))
+		if err != nil {
+			return err
+		}
+		err = h.Connect(ctx, pis[0])
+		if err != nil {
+			return err
+		}
+		h.SetStreamHandler(hello.ProtocolID, HandleStream)
+		weight, err := cs.Weight(ctx, ts)
+		if err != nil {
+			return err
+		}
+		gen := ts.Blocks()[0]
+		hmsg := &hello.HelloMessage{
+			HeaviestTipSet:       ts.Cids(),
+			HeaviestTipSetHeight: ts.Height(),
+			HeaviestTipSetWeight: weight,
+			GenesisHash:          gen.Cid(),
+		}
+		err = SayHello(ctx, pis[0].ID, h, hmsg)
+		if err != nil {
+			return err
+		}
+		time.Sleep(time.Second * 5)
+		return nil
+	},
+}
+
+func addrInfoFromArg(ctx context.Context, cctx *cli.Context) ([]peer.AddrInfo, error) {
```

Comment:
We should just make this function public in the CLI package and call it instead of duplicating code - https://github.com/filecoin-project/lotus/blob/master/cli/net.go#L293

Reply:
> We should just make this function public in the CLI package and call it instead of duplicating code - https://github.com/filecoin-project/lotus/blob/master/cli/net.go#L293

I did this at first, but I was worried that I should not modify lotus. I will submitted it soon